src/mlang/backend_ir/bir.ml

```diff
@@ -34,6 +34,39 @@ type program = {
   outputs : unit Mir.VariableMap.t;
 }
 
+let squish_statements (program : program) (threshold : int) (rule_suffix : string) =
+  let rule_from_stmts stmts =
+    let id = Mir.fresh_rule_id () in
+    { rule_id = id; rule_name = rule_suffix ^ string_of_int id; rule_stmts = List.rev stmts }
+  in
+  let rec browse_bir old_stmts new_stmts curr_stmts rules =
+    match old_stmts with
+    | [] -> (rules, List.rev (curr_stmts @ new_stmts))
+    | hd :: tl ->
+        let give_pos stmt = Pos.same_pos_as stmt hd in
+        let rules, curr_stmts =
+          match Pos.unmark hd with
+          | SConditional (expr, t, f) ->
+              let t_rules, t_curr_list = browse_bir t [] [] rules in
+              let f_rules, f_curr_list = browse_bir f [] [] t_rules in
+              let cond = give_pos (SConditional (expr, t_curr_list, f_curr_list)) in
+              (f_rules, cond :: curr_stmts)
+          | _ -> (rules, hd :: curr_stmts)
+        in
+        (* TODO: The following line works because of a low threshold. In order to make this robust,
```

Comment:
`List.length (List.flatten curr_stmts)` ?

Reply:
Unfortunately `curr_stmts` isn't a `'a list list` . Testing in utop and directly in the code doesn't seem to want to work. I think the only choice here is to recursively descend into the list, counting the lists inside `SConditional`s as we go.

Maybe I'm missing something?

Reply:
I found the solution with 10cbf45, you can take a look at it :)